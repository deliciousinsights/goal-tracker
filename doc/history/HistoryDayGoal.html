<!DOCTYPE html><html lang="en"><head><title>history/HistoryDayGoal</title></head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0"><meta name="groc-relative-root" content="../"><meta name="groc-document-path" content="history/HistoryDayGoal"><meta name="groc-project-path" content="src/history/HistoryDayGoal.js"><link rel="stylesheet" type="text/css" media="all" href="../assets/style.css"><script type="text/javascript" src="../assets/behavior.js"></script><body><div id="meta"><div class="file-path">src/history/HistoryDayGoal.js</div></div><div id="document"><div class="segment"><div class="comments "><div class="wrapper"><h1 id="journe-historique-dun-objectif">Journée historique d’un objectif</h1></div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>Sous-section de l&#39;écran d’historique, dédiée à un objectif sur un jour
précis.  Affiché à l&#39;intérieur d&#39;un autre composant : <code>HistoryDay</code>.</p></div></div><div class="code"><div class="wrapper"><span class="hljs-keyword">import</span> ListItem <span class="hljs-keyword">from</span> <span class="hljs-string">'@mui/material/ListItem'</span>
<span class="hljs-keyword">import</span> ListItemText <span class="hljs-keyword">from</span> <span class="hljs-string">'@mui/material/ListItemText'</span>

<span class="hljs-keyword">import</span> Gauge <span class="hljs-keyword">from</span> <span class="hljs-string">'../shared/Gauge'</span>
<span class="hljs-keyword">import</span> {
  GoalPropType,
  HistoryDayProgressesPropType,
} <span class="hljs-keyword">from</span> <span class="hljs-string">'../shared/prop-types'</span>

export default function HistoryDayGoal({</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>La déstructuration et les valeurs par défaut en force !</p></div></div><div class="code"><div class="wrapper">  goal: { name, units },
  stats: [progress = <span class="hljs-number">0</span>, target] = [],
}) {
  <span class="hljs-keyword">const</span> details = (
    <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-title">div</span>&gt;</span>
      <span class="hljs-tag">&lt;<span class="hljs-title">Gauge</span> <span class="hljs-attribute">value</span>=<span class="hljs-value">{progress}</span> <span class="hljs-attribute">max</span>=<span class="hljs-value">{target}</span> /&gt;</span>
      {progress} {units} sur {target}
    <span class="hljs-tag">&lt;/<span class="hljs-title">div</span>&gt;</span>
  )</span>

  <span class="hljs-keyword">return</span> (
    <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-title">ListItem</span>&gt;</span>
      <span class="hljs-tag">&lt;<span class="hljs-title">ListItemText</span>
        <span class="hljs-attribute">component</span>=<span class="hljs-value">'div'</span>
        <span class="hljs-attribute">primary</span>=<span class="hljs-value">{name}</span>
        <span class="hljs-attribute">secondary</span>=<span class="hljs-value">{details}</span>
        <span class="hljs-attribute">secondaryTypographyProps</span>=<span class="hljs-value">{{</span> <span class="hljs-attribute">component:</span> '<span class="hljs-attribute">div</span>' }}
      /&gt;</span>
    <span class="hljs-tag">&lt;/<span class="hljs-title">ListItem</span>&gt;</span>
  )</span>
}</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>Comme toujours, on définit les propriétés attendues/autorisées pour
validation.</p></div></div><div class="code"><div class="wrapper">HistoryDayGoal.propTypes = {
  goal: GoalPropType.isRequired,
  stats: HistoryDayProgressesPropType.isRequired,
}</div></div></div></div></body></html>