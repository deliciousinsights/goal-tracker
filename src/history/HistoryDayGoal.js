// Journée historique d’un objectif
// ================================

// Sous-section de l'écran d’historique, dédiée à un objectif sur un jour
// précis.  Affiché à l'intérieur d'un autre composant : `HistoryDay`.

import ListItem from '@mui/material/ListItem'
import ListItemText from '@mui/material/ListItemText'

import Gauge from '../shared/Gauge'
import {
  GoalPropType,
  HistoryDayProgressesPropType,
} from '../shared/prop-types'

export default function HistoryDayGoal({
  // La déstructuration et les valeurs par défaut en force !
  goal: { name, units },
  stats: [progress = 0, target] = [],
}) {
  const details = (
    <div>
      <Gauge value={progress} max={target} />
      {progress} {units} sur {target}
    </div>
  )

  return (
    <ListItem>
      <ListItemText
        component='div'
        primary={name}
        secondary={details}
        secondaryTypographyProps={{ component: 'div' }}
      />
    </ListItem>
  )
}

// Comme toujours, on définit les propriétés attendues/autorisées pour
// validation.
HistoryDayGoal.propTypes = {
  goal: GoalPropType.isRequired,
  stats: HistoryDayProgressesPropType.isRequired,
}
