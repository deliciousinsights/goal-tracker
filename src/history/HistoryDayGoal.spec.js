import { render } from '@testing-library/react'

import HistoryDayGoal from './HistoryDayGoal'

describe('<HistoryDayGoal />', () => {
  const goal = { id: 'abcd', name: 'Demo goal', target: 5, units: 'wombats' }

  it('should use proper stats propTypes', () => {
    checkFailPropType('stats in HistoryDayGoal must be an array.', () => (
      <HistoryDayGoal goal={goal} />
    ))
    checkFailPropType(
      'stats in HistoryDayGoal must be a pair of integers.',
      () => <HistoryDayGoal goal={goal} stats={[42]} />
    )
    checkFailPropType(
      'stats in HistoryDayGoal must start with a non-negative progress value.',
      () => <HistoryDayGoal goal={goal} stats={[-1, 10]} />
    )
    checkFailPropType(
      'stats in HistoryDayGoal must end with a positive target value.',
      () => <HistoryDayGoal goal={goal} stats={[0, 0]} />
    )
  })

  it('should otherwise match the expected snapshot', () => {
    const { container } = render(<HistoryDayGoal goal={goal} stats={[2, 5]} />)

    expect(container).toMatchSnapshot()
  })
})

function checkFailPropType(message, block) {
  const cErr = jest.spyOn(global.console, 'error').mockImplementation(() => {})
  try {
    block()
    expect(cErr).toHaveBeenCalledWith(
      'Warning: Failed %s type: %s%s',
      'prop',
      expect.stringMatching(
        new RegExp(message.replace(/[\].*+?(){}[]/g, '\\$&'))
      ),
      expect.anything()
    )
  } finally {
    cErr.mockRestore()
  }
}
