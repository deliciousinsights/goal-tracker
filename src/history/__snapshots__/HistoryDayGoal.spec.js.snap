// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HistoryDayGoal /> should otherwise match the expected snapshot 1`] = `
<div>
  <li
    class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
  >
    <div
      class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
      component="div"
    >
      <span
        class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
      >
        Demo goal
      </span>
      <div
        class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
      >
        <div>
          <span
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuenow="40"
            class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-eglki6-MuiLinearProgress-root"
            role="progressbar"
            style="height: 8px;"
          >
            <span
              class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
              style="transform: translateX(-60%);"
            />
          </span>
          2
           
          wombats
           sur 
          5
        </div>
      </div>
    </div>
  </li>
</div>
`;
